src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -325,15 +325,131 @@ static void uct_cuda_copy_sync_memops(CUdeviceptr dptr, int is_vmm)
                                   CU_POINTER_ATTRIBUTE_SYNC_MEMOPS, dptr));
 }
 
+static ucs_status_t
+uct_cuda_copy_push_ctx(uct_cuda_copy_md_t *md, CUdevice device,
+                       ucs_log_level_t log_level)
+{
+    ucs_status_t status;
+    CUcontext primary_ctx;
+
+    status = uct_cuda_primary_ctx_retain(
+                                     device,
+                                     md->config.retain_inactive_primary_ctx,
+                                     &primary_ctx);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = UCT_CUDADRV_FUNC(cuCtxPushCurrent(primary_ctx), log_level);
+    if (status != UCS_OK) {
+        (void)UCT_CUDADRV_FUNC(cuDevicePrimaryCtxRelease(device), log_level);
+    }
+
+    return status;
+}
+
+static void uct_cuda_copy_pop_ctx(CUdevice orig_device, CUdevice device,
+                                  ucs_log_level_t log_level)
+{
+    CUcontext ctx;
+    ucs_status_t status;
+
+    if (orig_device != device) {
+        status = UCT_CUDADRV_FUNC(cuCtxPopCurrent(&ctx), log_level);
+        if (status != UCS_OK) {
+            return;
+        }
+
+        status = UCT_CUDADRV_FUNC(cuDevicePrimaryCtxRelease(device), log_level);
+        if (status != UCS_OK) {
+            return;
+        }
```

Comment:
```suggestion
        UCT_CUDADRV_FUNC(cuCtxPopCurrent(NULL), log_level);
        UCT_CUDADRV_FUNC(cuDevicePrimaryCtxRelease(device), log_level);
```

No need to exit if pop failed. We may warn user about the error, and it is better to try release the primary context on the device anyway.

Reply:
fixed